docs/sources/mimir/manage/mimir-runbooks/_index.md

```diff
@@ -970,6 +970,27 @@ How to **investigate**:
        $.apps.v1.statefulSet.spec.template.metadata.withLabelsMixin({ [$._config.gossip_member_label]: 'false' }),
      ```
 
+### MimirKubernetesServiceEndpointsApproachingLimit
+
+This alert fires if a Kubernetes service used by Mimir for service discovery is approaching the endpoints limit.
+
+See [MimirKubernetesServiceEndpointsOverLimit](#MimirKubernetesServiceEndpointsOverLimit).
+
+### MimirKubernetesServiceEndpointsOverLimit
+
+This alert fires if a Kubernetes service used by Mimir for service discovery has reached the endpoints limit.
+
+How it **works**:
+
+- By default, the Kubernetes control plane creates and manages service endpoints to have no more than 100 endpoints each ([doc](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/)). You can configure this limit with the `--max-endpoints-per-slice kube-controller-manager` flag, up to a maximum of 1000.
+- Once a Kubernetes service endpoints list reached the limit, the endpoints are truncated to the limit.
```

Comment:
Is this how `EndpointSlices` work really? The way I thought about the endpoint-slices since reading the [announcement][1] was that

1. the control plane creates endpoint-slices per service
2. (by default) one endpoint-slice can include up to 100 endpoints (pointers to a pod)
3. when the limit is reached, the control plane creates additional endpoint-slices
4. since then the expectation was that a service can scale to beyond 100K endpoints.

Refer to the original "[Scaling Kubernetes Networking With EndpointSlices][1]" post.

Is this the problem we've encountered somewhere?

[1]: https://kubernetes.io/blog/2020/09/02/scaling-kubernetes-networking-with-endpointslices/

Reply:
The problem is that the DNS only returns the first 1K, at least that's what we observed. So when you do DNS-based service discovery, you're limited to the first 1K.

> Is this the problem we've encountered somewhere?

It's related to another issue we've observed. I can share more details offline about it.

Reply:
Does it make sense to check if individual endpoint slice has reached 100 limit, if DNS entry can return endpoints from more than 1 slice? Perhaps this alerts needs more investigation before merging.

Reply:
Out of the curiosity, I tried to catch up with the Kubernetes details, and I feel the "How it works" text is confusing here. I don't think there should be a connection between the `--max-endpoints-per-slice kube-controller-manager` and the observed DNS issue. The "1000 entries" problem feels like a coincidence rather than a relation to the `EndpointSlices`. Could it be so that the truncation happened on the client's resolver level (see also [Kubernetes DNS at scale][1])?

[1]: https://gist.github.com/aojea/32aeaa86aacebcdd93596ecb70fcba4f